google_guest_agent/instance_setup.go

```diff
@@ -193,6 +241,307 @@ func agentInit(ctx context.Context) {
 	}
 }
 
+func setMultiQueue(totalCPUs int) error {
+	devices, err := filepath.Glob(virtioNetDevs)
+	if err != nil {
+		return err
+	}
+	for _, dev := range devices {
+		if err := enableMultiQueue(dev); err != nil {
+			logger.Warningf("Could not enable multi queue for %s.", dev)
+			return err
+		}
+		if err = setQueueNumForDevice(dev); err != nil {
+			logger.Warningf("Could not set queue num for %s.", dev)
+			return err
+		}
+	}
+	// Set smp_affinity properly for gvnic queues. '-ntfy-block.' is unique to gve and will not affect virtio queues.
+	if err = setSMPAffinityForGVNIC(totalCPUs); err != nil {
+		logger.Warningf("Could not set smp_affinity for gvnic.")
+		return err
+	}
+	return nil
+}
+
+func setSMPAffinityForGVNIC(totalCPUs int) error {
+	irqDirs, err := filepath.Glob(irqDirPath)
+	if err != nil {
+		return err
+	}
+	for _, irq := range irqDirs {
+		blocks, err := filepath.Glob(irq + "/*-ntfy-block.*")
+		if err != nil {
+			return err
+		}
+		if len(blocks) > 0 && isFile(irq+"/affinity_hint") {
+			if err = copyFile(irq+"/affinity_hint", irq+"/smp_affinity"); err != nil {
+				return err
+			}
+		}
+	}
+	XPS, err := filepath.Glob(xpsCPU)
+	if err != nil {
+		return err
+	}
+
+	// If we have more CPUs than queues, then stripe CPUs across tx affinity as CPUNumber % queue_count.
+	for _, q := range XPS {
+		numQueues := len(XPS)
+		if numQueues > 63 {
+			numQueues = 63
+		}
+		r, _ := regexp.Compile(queueRegex)
+		match := r.MatchString(q)
+		var queueNum int
+		if match {
+			queueNum, err = strconv.Atoi(r.FindAllStringSubmatch(q, -1)[0][1])
+			if err != nil {
+				return err
+			}
+		}
+		xps := 0
+		for _, cpu := range makeRange(queueNum, numQueues, totalCPUs-1) {
+			xps |= 1 << cpu
+		}
+		// Linux xps_cpus requires a hex number with commas every 32 bits. It ignores
+		// all bits above # cpus, so write a list of comma separated 32 bit hex values
+		// with a comma between dwords.
+		var xpsDwords []string
+		for range makeRange(0, 1, (totalCPUs-1)/32) {
+			xpsDwords = append(xpsDwords, fmt.Sprintf("%08x", xps&0xffffffff))
+		}
+		var xpsString = strings.Join(xpsDwords, ",")
```

Comment:
can we get the two loops extracted to a function that just makes the xps string, and add some unit test for it?

Reply:
Done